src/qibo/parameter.py

```diff
@@ -0,0 +1,158 @@
+
+        if (self.nfeat + self.nparams) != len(parameters):
+            raise_error(
+                ValueError,
+                f"The lambda function has {len(parameters)} parameters, the input has {self.nfeat+self.nparams}.",
+            )
+
+        iterator = iter(parameters.items())
+
+        for i in range(self.nfeat):
+            x = next(iterator)
+            if x[0][0] != "x":
+                raise_error(
+                    ValueError,
+                    f"Parameter #{i} in the lambda function should be a feature starting with `x`",
+                )
+
+        for i in range(self.nparams):
+            x = next(iterator)
+            if x[0][:2] != "th":
+                raise_error(
+                    ValueError,
+                    f"Parameter #{self.nfeat+i} in the lambda function should be a trainable parameter starting with `th`",
+                )
+
+    def _apply_func(self, function, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
+        params = []
+        if self._feature is not None:
+            if isinstance(self._feature, list):
+                params.extend(self._feature)
+            else:
+                params.append(self._feature)
+        if fixed_params:
+            params.extend(fixed_params)
+        else:
+            params.extend(self._trainable)
+        return float(function(*params))
+
+    def _calculate_derivatives(self):
+        """Calculates derivatives w.r.t to all trainable parameters"""
+        vars = []
+        for i in range(self.nfeat):
+            vars.append(sp.Symbol(f"x{i}"))
+        for i in range(self.nparams):
+            vars.append(sp.Symbol(f"th{i}"))
+
+        expr = sp.sympify(self.lambdaf(*vars))
+
+        derivatives = []
+        for i in range(len(vars)):
+            derivative_expr = sp.diff(expr, vars[i])
+            derivatives.append(sp.lambdify(vars, derivative_expr))
+
+        return derivatives
+
+    def update_parameters(self, trainable=None, feature=None):
+        """Update gate trainable parameter and feature values"""
+        if not isinstance(trainable, (list, np.ndarray)):
+            raise_error(
+                ValueError, "Trainable parameters must be given as list or numpy array"
+            )
+
+        if self.nparams != len(trainable):
+            raise_error(
+                ValueError,
+                f"{len(trainable)} trainable parameters given, need {self.nparams}",
+            )
+
+        if not isinstance(feature, (list, np.ndarray)) and self._feature != feature:
+            raise_error(ValueError, "Features must be given as list or numpy array")
+
+        if self._feature is not None and self.nfeat != len(feature):
+            raise_error(ValueError, f"{len(feature)} features given, need {self.nfeat}")
+
+        if trainable is not None:
+            self._trainable = trainable
+        if feature and self._feature:
```

Comment:
Don't test for falsiness, but for their actual value, i.e.
```suggestion
        if feature is not None and self._feature is not None:
```